pkg/xipfs/fs/xipfs_fs.c

```diff
@@ -0,0 +1,804 @@
+}
+
+static ssize_t _read(vfs_file_t *filp, void *dest, size_t nbytes)
+{
+    xipfs_file_desc_t *descp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(filp->mp);
+    descp = (xipfs_file_desc_t *)(uintptr_t)&filp->private_data.ptr;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_read(mp, descp, dest, nbytes);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+static ssize_t _write(vfs_file_t *filp, const void *src, size_t nbytes)
+{
+    xipfs_file_desc_t *descp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(filp->mp);
+    descp = (xipfs_file_desc_t *)(uintptr_t)&filp->private_data.ptr;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_write(mp, descp, src, nbytes);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+static int _fsync(vfs_file_t *filp)
+{
+    xipfs_file_desc_t *descp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(filp->mp);
+    descp = (xipfs_file_desc_t *)(uintptr_t)&filp->private_data.ptr;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_fsync(mp, descp, filp->pos);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+/*
+ * Operations on open directories
+ */
+
+static int _opendir(vfs_DIR *dirp, const char *dirname)
+{
+    xipfs_dir_desc_t *descp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(dirp->mp);
+    descp = (xipfs_dir_desc_t *)(uintptr_t)&dirp->private_data.ptr;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_opendir(mp, descp, dirname);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+static int _readdir(vfs_DIR *dirp, vfs_dirent_t *entry)
+{
+    xipfs_dir_desc_t *descp;
+    xipfs_dirent_t *direntp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(dirp->mp);
+    descp = (xipfs_dir_desc_t *)(uintptr_t)&dirp->private_data.ptr;
+    direntp = (xipfs_dirent_t *)(uintptr_t)entry->d_name;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_readdir(mp, descp, direntp);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+static int _closedir(vfs_DIR *dirp)
+{
+    xipfs_dir_desc_t *descp;
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(dirp->mp);
+    descp = (xipfs_dir_desc_t *)(uintptr_t)&dirp->private_data.ptr;
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_closedir(mp, descp);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+/*
+ * Operations on mounted file systems
+ */
+
+static int _format(vfs_mount_t *vfs_mp)
+{
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(vfs_mp);
+
+    mutex_lock(mp->execution_mutex);
+    mutex_lock(mp->mutex);
+    ret = xipfs_format(mp);
+    mutex_lock(mp->mutex);
+    mutex_lock(mp->execution_mutex);
+
+    return ret;
+}
+
+static int _mount(vfs_mount_t *vfs_mp)
+{
+    xipfs_mount_t *mp;
+    int ret;
+
+    mp = _get_xipfs_mount_t(vfs_mp);
+
+    mutex_lock(mp->mutex);
+    ret = xipfs_mount(mp);
+    mutex_unlock(mp->mutex);
+
+    return ret;
+}
+
+static int _umount(vfs_mount_t *vfs_mp)
+{
+    UNUSED(vfs_mp);
```

Comment:
```suggestion
    (void)vfs_mp;
```

The `UNUSED` macro is only used once, so it's not really worth defining a macro for that. Also, typically in RIOT we use the `(void)variable_name;` variant.